src/python/pants/option/arg_splitter_test.py

```diff
@@ -98,140 +86,259 @@ def assert_not_positional(arg):
-    self._split('./pants test ./test', ['test'], {'': [], 'test': []}, ['./test'])
-    self._split('./pants test //test', ['test'], {'': [], 'test': []}, ['//test'])
-
-  def test_descoping_qualified_flags(self):
-    self._split('./pants compile test --compile-java-bar --no-test-junit-baz foo/bar',
-                ['compile', 'test'],
-                {'': [], 'compile': [], 'compile.java': ['--bar'], 'test': [],
-                 'test.junit': ['--no-baz']}, ['foo/bar'])
+    self.assert_valid_split(
+      './pants --compile-java-long-flag -f compile -g compile.java -x test.junit -i '
+      'src/java/org/pantsbuild/foo src/java/org/pantsbuild/bar:baz',
+      expected_goals=['compile', 'test'],
+      expected_scope_to_flags={
+        '': ['-f'],
+        'compile.java': ['--long-flag', '-x'],
+        'compile': ['-g'],
+        'test.junit': ['-i']
+      },
+      expected_positional_args=['src/java/org/pantsbuild/foo', 'src/java/org/pantsbuild/bar:baz'],
+    )
+    self.assert_valid_split(
+      './pants -farg --fff=arg compile --gg-gg=arg-arg -g test.junit --iii '
+      '--compile-java-long-flag src/java/org/pantsbuild/foo src/java/org/pantsbuild/bar:baz',
+      expected_goals=['compile', 'test'],
+      expected_scope_to_flags={
+        '': ['-farg', '--fff=arg'],
+        'compile': ['--gg-gg=arg-arg', '-g'],
+        'test.junit': ['--iii'],
+        'compile.java': ['--long-flag'],
+      },
+      expected_positional_args=['src/java/org/pantsbuild/foo', 'src/java/org/pantsbuild/bar:baz'],
+    )
+
+  def test_distinguish_goals_from_positional_args(self) -> None:
+    self.assert_valid_split(
+      './pants compile test foo::',
+      expected_goals=['compile', 'test'],
+      expected_scope_to_flags={'': [], 'compile': [], 'test': []},
+      expected_positional_args=['foo::'],
+    )
+    self.assert_valid_split(
+      './pants compile test foo::',
+      expected_goals=['compile', 'test'],
+      expected_scope_to_flags={'': [], 'compile': [], 'test': []},
+      expected_positional_args=['foo::'],
+    )
+    self.assert_valid_split(
+      './pants compile test:test',
+      expected_goals=['compile'],
+      expected_scope_to_flags={'': [], 'compile': []},
+      expected_positional_args=['test:test'],
+    )
+    self.assert_valid_split(
+      './pants test test:test',
+      expected_goals=['test'],
+      expected_scope_to_flags={'': [], 'test': []},
+      expected_positional_args=['test:test'],
+    )
+    self.assert_valid_split(
+      './pants test ./test',
+      expected_goals=['test'],
+      expected_scope_to_flags={'': [], 'test': []},
+      expected_positional_args=['./test'],
+    )
+    self.assert_valid_split(
+      './pants test //test',
+      expected_goals=['test'],
+      expected_scope_to_flags={'': [], 'test': []},
+      expected_positional_args=['//test'],
+    )
 
+  def test_descoping_qualified_flags(self) -> None:
```

Comment:
This function is the same, modulo using keyword args and some formatting tweaks.